rmc/utils/mpc.py

```diff
@@ -669,158 +857,177 @@ def aggregate_gnomad_fitted_scores(
-    )
 
-    logger.info("Creating MPC release and writing out...")
-    ht = ht.annotate(mpc=-(hl.log10(ht.n_less / gnomad_var_count)))
+        # Search all gnomAD scores to find first score that is
+        # greater than or equal to score to be annotated
+        # `binary_search` will return the index of the first gnomAD fitted score that
+        # is >= the score of interest
+        # e.g., if the score of interest is 0.45, and gnomAD fitted scores are
+        # [0.3, 0.4, 0.5], then `binary_search` will return an index of 2
+        # the `n_less` of 0.5 will contain the counts of variants with gnomAD scores of
+        # 0.3 and 0.4 due to the non-inclusive nature of scans
+        # (n_less[0.5] = n_var[0.3] + n_var[0.4])
+        scores_ht = scores_ht.annotate(
+            idx=hl.binary_search(scores_ht.gnomad_scores, scores_ht.fitted_score)
+        )
+        scores_ht = scores_ht.annotate(
+            n_less=hl.if_else(
+                # Make n_less equal to total gnomAD common variant count if
+                # index is equal to the length of the gnomAD scores array
+                scores_ht.idx == gnomad_scores_len,
+                gnomad_var_count,
+                gnomad_ht[scores_ht.idx].n_less,
+            )
+        )
+        # Checkpoint here to force the binary search to compute
+        scores_ht = scores_ht.checkpoint(
+            f"{TEMP_PATH_WITH_FAST_DEL}/mpc_temp_binary{temp_label}.ht",
+            _read_if_exists=not overwrite_temp,
+            overwrite=overwrite_temp,
+        )
 
-    ht = ht.checkpoint(
-        mpc_release.versions[freeze].path,
-        overwrite=overwrite_output,
-        _read_if_exists=not overwrite_output,
-    )
-    logger.info("Output MPC HT fields: %s", set(ht.row))
+        logger.info("Calculating MPC scores...")
+        scores_ht = scores_ht.annotate(
+            mpc=-(hl.log10(scores_ht.n_less / gnomad_var_count))
+        )
+        scores_ht = scores_ht.checkpoint(
+            f"{TEMP_PATH_WITH_FAST_DEL}/mpc{temp_label}.ht",
+            _read_if_exists=not overwrite_temp,
+            overwrite=overwrite_temp,
+        )
 
-    logger.info("Creating dedup MPC release and writing out...")
-    ht = ht.select("mpc").key_by("locus", "alleles")
-    ht = ht.collect_by_key()
-    ht = ht.annotate(**{"mpc": hl.nanmax(ht.values["mpc"])})
-    ht = ht.annotate(
-        **{"transcript": ht.values.find(lambda x: x["mpc"] == ht["mpc"]).transcript}
-    )
-    ht = ht.drop("values")
-    ht.write(mpc_release_dedup.versions[freeze].path, overwrite=overwrite_output)
-    logger.info("Output MPC dedup HT fields: %s", set(ht.row))
+        logger.info("Annotating MPC scores to input table and writing out...")
+        ht = ht.annotate(mpc=scores_ht[ht.key].mpc)
+        ht.write(output_ht_path, overwrite=True)
 
 
-def annotate_mpc(
-    ht: hl.Table,
-    output_path: str,
-    overwrite: bool = True,
-    add_transcript_annotation: bool = True,
+def create_mpc_release_ht(
```

Comment:
maybe we should remove this function entirely after your reorganization of `annotate_mpc`? this part
```
    annotate_mpc(
        ht=context_with_oe.versions[freeze].ht().select(),
        output_ht_path=mpc_release.versions[freeze].path,
        temp_label="_release",
        use_release=False,
        overwrite_temp=overwrite_temp,
        freeze=freeze,
    )
```
could get called in `calculate_mpc.py`